### PR TITLE
SIQ-531 Add functionality to authorize on scope

### PR DESCRIPTION
**Pull request by ryan-carlson**
**Issue Link**
https://spothero.atlassian.net/browse/SIQ-531

**Description**
This change will allow authorization against a slice of required scopes.  Each of the required scopes must exist in the scope taken from the token to complete authorization.

### CODE REVIEW

**Review by theisenmark — APPROVED** [review]
A couple suggestions, but nothing major. Nice job!

File: jose/middleware.go [hunk 1]
```diff
@@ -151,29 +153,73 @@ func AuthenticationMiddleware(next http.Handler) http.Handler {
 	return EnforceAuthentication(next.ServeHTTP)
 }
 
-// EnforceAuthentication enforces authentication for a single HTTP handler.
-func EnforceAuthentication(next http.HandlerFunc) http.HandlerFunc {
+type AuthParams struct {
+	requiredScopes []string
+}
+
+func validateRequiredScope(r *http.Request, params AuthParams) (valid bool, message string) {
```
[the remark this answers is not in the document]
Comment (ryan-carlson) [reply jose/middleware.go]: Updated

File: jose/middleware.go [hunk 2]
```diff
@@ -151,29 +153,73 @@ func AuthenticationMiddleware(next http.Handler) http.Handler {
...
+	if params.requiredScopes != nil {
+		claim, err := FromContext(r.Context())
+		if err != nil {
+			return false, cannotFindClaim
+		}
+
+		scope := strings.Split(claim.Scope, " ")
```
[the remark this answers is not in the document]
Comment (ryan-carlson) [reply jose/middleware.go]: Updated

File: jose/middleware.go [hunk 3]
```diff
@@ -151,29 +153,73 @@ func AuthenticationMiddleware(next http.Handler) http.Handler {
...
+}
+
+func hasScope(requiredScope string, scope []string) bool {
+	for i := range scope {
+		if scope[i] == requiredScope {
+			return true
+		}
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (ryan-carlson) [reply jose/middleware.go]: I took a stab at this, see https://github.com/spothero/tools/pull/171/commits/0091ed4b721729674bb21cba6f51cfc6eaa329be

**Review by theisenmark — APPROVED** [review]

**Pull request merged**